File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
sys.setrecursionlimit(10000)
input = sys.stdin.readline

def dfs(x, y):
    visited[x][y] = True
    for dx, dy in [(-1,0),(1,0),(0,-1),(0,1)]:
        nx, ny = x + dx, y + dy
        if 0 <= nx < N and 0 <= ny < M and field[nx][ny] == 1 and not visited[nx][ny]:
            dfs(nx, ny)

T = int(input())
for _ in range(T):
    M, N, K = map(int, input().split())
    field = [[0] * M for _ in range(N)]
    visited = [[False] * M for _ in range(N)]
    
    for _ in range(K):
        x, y = map(int, input().split())
        field[y][x] = 1
    
    count = 0
    for i in range(N):
        for j in range(M):
            if field[i][j] == 1 and not visited[i][j]:
                dfs(i, j)
                count += 1
    
    print(count)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const input = require('fs').readFileSync(0).toString().trim().split('\n');
let idx = 0;

const T = Number(input[idx++]);
for (let t = 0; t < T; t++) {
    const [M, N, K] = input[idx++].split(' ').map(Number);
    const field = Array.from({length: N}, () => new Array(M).fill(0));
    const visited = Array.from({length: N}, () => new Array(M).fill(false));

    for (let i = 0; i < K; i++) {
        const [x, y] = input[idx++].split(' ').map(Number);
        field[y][x] = 1;
    }

    function dfs(x, y) {
        const stack = [[x, y]];
        visited[x][y] = true;
        while (stack.length > 0) {
            const [cx, cy] = stack.pop();
            for (const [dx, dy] of [[-1,0],[1,0],[0,-1],[0,1]]) {
                const nx = cx + dx;
                const ny = cy + dy;
                if (0 <= nx && nx < N && 0 <= ny && ny < M && field[nx][ny] === 1 && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    stack.push([nx, ny]);
                }
            }
        }
    }

    let count = 0;
    for (let i = 0; i < N; i++) {
        for (let j = 0; j < M; j++) {
            if (field[i][j] === 1 && !visited[i][j]) {
                dfs(i, j);
                count++;
            }
        }
    }
    console.log(count);
}